### ISSUE TITLE
DuckDuckGPT safari

### ISSUE THREAD

**Issue by Delamcode** [adamlui/userscripts#10]
I know there is not official support, but using the userscripts extension I get this error: GM_getValue is not a function. (In 'GM_getValue("accessToken")', 'GM_getValue' is undefined). Do you have an idea of what is causing this? Here is the extension readme: https://github.com/quoid/userscripts

**Comment by adamlui** [comment]
Hey, I don't have a Mac so can't test support but can you try it using Tampermonkey @ https://www.tampermonkey.net/index.php?browser=safari&locale=en

**Comment by Delamcode** [comment]
No, unfortunately I can’t. I was just wondering if it is something that would be easy to fix…

**Comment by adamlui** [comment]
How come? If you could it would probably work, GM_getValue is a function supported by userscript managers like Tampermonkey & Violentmonkey, the readme you linked is for a userscript editor only

**Comment by Delamcode** [comment]
Because only userscripts supports ipad. It is not only an editor. Here is the popup in safari: https://github.com/quoid/userscripts#popup

**Comment by adamlui** [comment]
Ohh, well I asked da bot and it said:

![image](https://user-images.githubusercontent.com/10906554/219954610-65a9d314-0e97-417f-9f8c-fbe8dc85c9de.png)

Do you know how to edit the code? Can you remove line 17:

`// @grant            GM_getValue`

Remove line 29:

`var GM_getValue = (() => window.GM_getValue)()`

And finally, replace line 168 with:

`var accessToken = localStorage.getItem("accessToken")`

...and let me know if it works?

**Comment by Delamcode** [comment]
Weird… the readme says it is supported but I’ll try and report back

**Comment by adamlui** [comment]
Ok then

**Comment by adamlui** [comment]
Actually you could try asking the extension author why the function wasn't supported on your iPad, and maybe provide them screenshots

**Comment by Delamcode** [comment]
Not sure how to do that. I now get this error: undefined is not a function (near '...GM_xmlhttpRequest...')
asyncFunctionResume@[native code]
@[native code]
promiseReactionJobWithoutPromise@[native code]
appendChild@[native code]
injectJS@safari-web-extension://F0435466-F9A8-46C1-B498-8953D9410A63/content.js:274:34
processJS@safari-web-extension://F0435466-F9A8-46C1-B498-8953D9410A63/content.js:222:21
@safari-web-extension://F0435466-F9A8-46C1-B498-8953D9410A63/content.js:402:18

**Comment by Delamcode** [comment]
> Actually you could try asking the extension author why the function wasn't supported on your iPad, and maybe provide them screenshots + link to my code

Sure, I could try

**Comment by Delamcode** [comment]
Is there any setup I’m missing?

**Comment by adamlui** [comment]
> I now get this error: undefined is not a function (near '...GM_xmlhttpRequest...') asyncFunctionResume@[native code] @[native code] promiseReactionJobWithoutPromise@[native code] appendChild@[native code] injectJS@safari-web-extension://F0435466-F9A8-46C1-B498-8953D9410A63/content.js:274:34 processJS@safari-web-extension://F0435466-F9A8-46C1-B498-8953D9410A63/content.js:222:21 @safari-web-extension://F0435466-F9A8-46C1-B498-8953D9410A63/content.js:402:18

Yeah it looks like none of the GM_ functions are supported by that extension & DuckDuckGPT uses these 5: 

`// @grant            GM_deleteValue`
`// @grant            GM_getValue`
`// @grant            GM_setValue`
`// @grant            GM_info`
`// @grant            GM_xmlhttpRequest`

**Comment by Delamcode** [comment]
Interesting… I wonder why they don’t work as [here](https://github.com/quoid/userscripts#api) it claims it’s supported…

**Comment by Delamcode** [comment]
Could it be that userscripts uses GM.[Item] instead of GM_[Item]?

**Comment by Delamcode** [comment]
Opened an issue on the extension ^^

**Comment by Delamcode** [comment]
Maybe it has something to do with [this](https://github.com/quoid/userscripts/issues/265)?

**Comment by adamlui** [comment]
> Maybe it has something to do with [this](https://github.com/quoid/userscripts/issues/265)?

I'ma look at it tonight after I sleep (I was up all night)

**Comment by Delamcode** [comment]
According to the [other issue](https://github.com/quoid/userscripts/issues/423) I should change it to GM. but then I get this error (will edit if anything changes): 
`ReferenceError {} column: 43 line: 87 message: "Can't find variable: GM_registerMenuCommand" sourceURL: "https://duckduckgo.com/?q=what+is+the+meaning+of+life&t=ipad" stack: "enableCommandMenu@https://duckduckgo.com/?q=what+is+the+meaning+of+life&t=ipad:87:43↵main@https://duckduckgo.com/?q=what+is+the+meaning+of+life&t=ipad:54:26↵@https://duckduckgo.com/?q=what+is+the+meaning+of+life&t=ipad:112:17↵@https://duckduckgo.com/?q=what+is+the+meaning+of+life&t=ipad:179:3↵@https://duckduckgo.com/?q=what+is+the+meaning+of+life&t=ipad:181:15↵global code@https://duckduckgo.com/?q=what+is+the+meaning+of+life&t=ipad:182:11↵appendChild@[native code]↵injectJS@safari-web-extension://F0435466-F9A8-46C1-B498-8953D9410A63/content.js:274:34↵processJS@safari-web-extension://F0435466-F9A8-46C1-B498-8953D9410A63/content.js:222:21↵@safari-web-extension://F0435466-F9A8-46C1-B498-8953D9410A63/content.js:402:18" __proto__: Object `

**Comment by quoid** [comment]
I am one of the maintainers of the previously mentioned [user script manager](https://github.com/quoid/userscripts). The reason an error is thrown is because synchronous `GM_` apis are not supported by the manager (outside of `GM_info` and `GM_xmlhttpRequest`). Only certain `async` counter parts are supported - [readme](https://github.com/quoid/userscripts#api). The main difference in these api methods being `sync` vs `async`.

I recommend the newer API methods, but thats obviously up to the user script author. If you'd like to support the newer API methods you could do something like:

```js
...
// @grant            GM_deleteValue
// @grant            GM_getValue
// @grant            GM_setValue
// @grant            GM_info
// @grant            GM_xmlhttpRequest
// @grant            GM.deleteValue
// @grant            GM.getValue
// @grant            GM.setValue
// @grant            GM.info
// @grant            GM.xmlHttpRequest
...
var GM_setValue = (() => window.GM_setValue ?? GM.setValue)()
var GM_deleteValue = (() => window.GM_deleteValue ?? GM.deleteValue)()
var GM_info = (() => window.GM_info ?? GM.info)()
var GM_xmlhttpRequest = (() => window.GM_xmlhttpRequest ?? GM.xmlHttpRequest)()
var GM_getValue = (() => window.GM_getValue ?? GM.getValue)()
...
```

**Comment by Delamcode** [comment]
Using just these updates makes it “work” but asks to log in even though I’m logged in…

**Comment by adamlui** [comment]
@Delamcode did you revert the edits in https://github.com/adamlui/userscripts/issues/10#issuecomment-1436006798 before applying @quoid's tweak? If so, I asked da bot and apparently the "nullish coalescing operator (`??`), which is a relatively new feature introduced in ECMAScript 2020 [is not always supported]. Unfortunately, not all browsers support this feature yet, including some versions of Safari on iPad."

Can you try replacing them with the `OR` operator?

`var GM_setValue = (() => window.GM_setValue || GM.setValue)()`
`var GM_setValue = (() => window.GM_setValue || GM.setValue)()`
`var GM_deleteValue = (() => window.GM_deleteValue || GM.deleteValue)()`
`var GM_info = (() => window.GM_info || GM.info)()`
`var GM_xmlhttpRequest = (() => window.GM_xmlhttpRequest || GM.xmlHttpRequest)()`
`var GM_getValue = (() => window.GM_getValue || GM.getValue)()`

Also what version of Safari are you using, is it 10+?

**Comment by Delamcode** [comment]
I am using safari 15. Yes, I did revert. I will try this.

**Comment by Delamcode** [comment]
Now both methods stopped working… (Waiting for Response) with this error: JSON Parse error: Unrecognized token '<'
onload@
@safari-web-extension://ED32EE2A-2F54-40AF-89E4-E1F764226C05/content.js:180:38
asyncFunctionResume@[native code]
@safari-web-extension://ED32EE2A-2F54-40AF-89E4-E1F764226C05/content.js:156:59

**Comment by adamlui** [comment]
Actually I just found out ChatGPT's been down forever...
![image](https://user-images.githubusercontent.com/10906554/220334053-da8d39fd-33dc-4d63-9ab1-1d29e3c772a5.png)
So can you go back to using the ??'s, I think it might work when they're back up

**Comment by Delamcode** [comment]
The actual website is working fine for me, and I am logged in. Using ?? I still get the log in @ error, even though I’m logged in…

**Comment by Delamcode** [comment]
Is it possible to console.log the id on my computer, and then hard code it on my ipad script?

**Comment by adamlui** [comment]
> Is it possible to console.log the id on my computer, and then hard code it on my ipad script?

Probably, also you could try adding 'GM_deleteValue("accessToken")' after line 27, save, refresh search page, then remove the line, re-save, then refresh search page (just to clear it out)

**Comment by Delamcode** [comment]
Where (and how) would you recommend doing the console.log?

**Comment by adamlui** [comment]
Can you try clearing the access token first see if that works? (I had to do this in incompatible browsers is why I didn't include them)

**Comment by Delamcode** [comment]
No it didn’t work. After declaring the var right? Because line 27 isn’t the same for us

**Comment by adamlui** [comment]
Can you add these 3 lines here
![image](https://user-images.githubusercontent.com/10906554/220367952-c2a0af06-c000-41dd-a7a9-b2d43256b39f.png)
then refresh DDG, then after message says it's cleared, remove the 3 lines, re-save the code, then re-fresh DDG

**Comment by Delamcode** [comment]
Now I get this error again: 
JSON Parse error: Unrecognized token '<'
onload@
@safari-web-extension://879A7AA8-7197-4877-8688-216230ABB051/content.js:180:38
asyncFunctionResume@[native code]
@safari-web-extension://879A7AA8-7197-4877-8688-216230ABB051/content.js:156:59

**Comment by Delamcode** [comment]
I also made a [fork](https://github.com/Delamcode/DuckDuckGPT-Safari-iOS/blob/master/chatgpt/duckduckgpt/duckduckgpt.user.js) here.

**Comment by adamlui** [comment]
How much are iPads? I think I need to buy one to make it work so I can trial & error without burning you out

**Comment by adamlui** [comment]
> Now I get this error again: JSON Parse error: Unrecognized token '<' onload@ @safari-web-extension://879A7AA8-7197-4877-8688-216230ABB051/content.js:180:38 asyncFunctionResume@[native code] @safari-web-extension://879A7AA8-7197-4877-8688-216230ABB051/content.js:156:59

Before, this was when you removed the ??'s, are they still in place?

**Comment by Delamcode** [comment]
> It’s not a problem, but you can get a last gen base ipad for $330 officially.

Also I feel that you should get a higher end one if you can (iPad Air 5 for 600) or get a used iPad for cheaper

**Comment by adamlui** [comment]
I did some research, when it returns '<' that means you are getting a HTML response like 404 or 500 error. Can you try adding this line that might fix it for iOS

![image](https://user-images.githubusercontent.com/10906554/220373755-20927923-38e8-44f9-84e2-b83526fccaa8.png)

**Comment by adamlui** [comment]
> > It’s not a problem, but you can get a last gen base ipad for $330 officially.
> 
> Also I feel that you should get a higher end one if you can (iPad Air 5 for 600) or get a used iPad for cheaper

If stuff works on lower end it's sure to work on higher too though

**Comment by Delamcode** [comment]
> > > It’s not a problem, but you can get a last gen base ipad for $330 officially.
> > 
> > 
> > Also I feel that you should get a higher end one if you can (iPad Air 5 for 600) or get a used iPad for cheaper
> 
> If stuff works on lower end it's sure to work on higher too though

Fair.

**Comment by Delamcode** [comment]
> I did some research, when it returns '<' that means you are getting a HTML response like 404 or 500 error. Can you try adding this line that might fix it for iOS
> 
> ![image](https://user-images.githubusercontent.com/10906554/220373755-20927923-38e8-44f9-84e2-b83526fccaa8.png)

Yeah that fixed it, but still getting that login error now…

**Comment by adamlui** [comment]
Can you Ctrl+F for `alertLogin()`, there are 4 occurrences, 1st is function declaration, skip that one & before the 2nd occurrence, can you add this `alert('Unauthorized')`:
``` js
...
        if (error === "UNAUTHORIZED") {
            GM_deleteValue("accessToken")
            alert('Unauthorized')
            alertLogin()
        }
...
```
Then in the line of 3rd occurrence, can you insert `alert('401') ;` before it: 
```js
...
    function onLoad() {
        return function(event) {
            if (event.status === 401) { GM_deleteValue("accessToken") ; alert('401') ; alertLogin() }
            if (event.status === 403) { alertBlockedByCloudflare() }
            if (event.status === 429) { alertFrequentRequests() }
            if (getUserscriptManager() !== "Tampermonkey") {
...
```
Then before the 4th occurrence, can you insert `alert('Cloudflare')` before it:
```
...
                    if (isBlockedbyCloudflare(response.responseText)) {
                        alert('Cloudflare')
                        alertLogin() ; return }
                    var accessToken2 = JSON.parse(response.responseText).accessToken;
...
```
...then tell me which alert appears?

**Comment by adamlui** [comment]
Actually you could use console.log instead so you don't have to delete them

**Comment by adamlui** [comment]
Do you like techno?

**Comment by Delamcode** [comment]
1. Working on it
2. Not sure what you mean 
3. No?

**Comment by adamlui** [comment]
Instead of alert() you can use console.log() if you can read it from iPad, then annoying pop-ups won't appear until you unedit the code (which you won't have to do in case keeping that logging proves useful)

**Comment by Delamcode** [comment]
Can you Ctrl+F for `alertLogin()`, there are 4 occurrences, 1st is function declaration, skip that one & before the 2nd occurrence, can you add this `alert('Unauthorized')`:
> 
> ```js
> ...
>         if (error === "UNAUTHORIZED") {
>             GM_deleteValue("accessToken")
>             alert('Unauthorized')
>             alertLogin()
>         }
> ...
> ```
> 
> Then in the line of 3rd occurrence, can you insert `alert('401') ;` before it:
> 
> ```js
> ...
>     function onLoad() {
>         return function(event) {
>             if (event.status === 401) { GM_deleteValue("accessToken") ; alert('401') ; alertLogin() }
>             if (event.status === 403) { alertBlockedByCloudflare() }
>             if (event.status === 429) { alertFrequentRequests() }
>             if (getUserscriptManager() !== "Tampermonkey") {
> ...
> ```
> 
> Then before the 4th occurrence, can you insert `alert('Cloudflare')` before it:
> 
> ```js
> ...
>                     if (isBlockedbyCloudflare(response.responseText)) {
>                         alert('Cloudflare')
>                         alertLogin() ; return }
>                     var accessToken2 = JSON.parse(response.responseText).accessToken;
> ...
> ```
> 
> ...then tell me which alert appears?

Sorry, that took a while:
I got the Unauthorized error

**Comment by Delamcode** [comment]
> Instead of alert() you can use console.log() if you can read it from iPad, then annoying pop-ups won't appear until you unedit the code (which you won't have to do in case keeping that logging proves useful)
Alright I’ll update that just in case

**Comment by adamlui** [comment]
> > Instead of alert() you can use console.log() if you can read it from iPad, then annoying pop-ups won't appear until you unedit the code (which you won't have to do in case keeping that logging proves useful)
> 
> Alright I’ll update that just in case

It's OK you can leave it since it's more conspicuous, now can you add `alert(accessToken)` here:
![image](https://user-images.githubusercontent.com/10906554/220389442-7508a9cf-892b-45a4-8e28-7b2910e9d099.png)
...and is the output really long with multiple lines of numbers and letters?

**Comment by Delamcode** [comment]
Unfortunately no, I just get undefined…

**Comment by Delamcode** [comment]
Sure I might look into it if I have time!

**Comment by quoid** [comment]
> Now I get this error again: JSON Parse error: Unrecognized token '<' ...
> that means you are getting a HTML response like 404 or 500 error...
> How much are iPads? I think I need to buy one to make it work ...

This is happening on macOS as well, not just iPadOS/iOS.

You can just log the response in `GM_xmlhttpRequest`, like so:

```js
...
onload: function(response) {
    console.log('r', response)
    ...
}
...
```

It's a `403` error, I pasted the response below:

```js
{
    readyState: 4
    response: "<html>↵  <head>↵    <style global>body{font-family:Arial,Helvetica,sans-serif}.container{align-items:center;display:flex;flex-direction:colu…"
    responseHeaders: "alt-svc: h3=\":443\"; ma=86400, h3-29=\":443\"; ma=86400\r↵cache-control: private, max-age=0, no-store, no-cache, must-revalidate, post-check=0…"
    responseText: "<html>↵  <head>↵    <style global>body{font-family:Arial,Helvetica,sans-serif}.container{align-items:center;display:flex;flex-direction:colu…"
    responseType: ""
    responseURL: "https://chat.openai.com/api/auth/session"
    status: 403
    statusText: "Forbidden"
    timeout: 0
}
```

If you add a check for that here:

```js
if (isBlockedbyCloudflare(response.responseText) || response.status === 403) {
   alertLogin(); return 
}
```

The `403` error will go away, but it looks like the subsequent `200` responses are returning an empty object, pasted below:

```js
{
    readyState: 4
    response: "{}"
    responseHeaders: "alt-svc: h3=\":443\"; ma=86400, h3-29=\":443\"; ma=86400\r↵cf-cache-status: DYNAMIC\r↵cf-ray: 79d39c8a29d5e6e4-EWR\r↵content-length: 2…"
    responseText: "{}"
    responseType: ""
    responseURL: "https://chat.openai.com/api/auth/session"
    status: 200
    statusText: "OK"
    timeout: 0
    withCredentials: false
}
```

I am not entirely sure how this user script is supposed to function, but it seems you are expecting an access token from that response. I am not sure the context where the request is run has access to that token. I believe it's akin to making the request in an incognito tab, so I am not sure how one would expect to the actual token.

I am not sure how to work around this.

**Comment by adamlui** [comment]
Hey @quoid the userscript forces user to visit login page if:

1. POST request to `https://chat.openai.com/backend-api/conversation` is unauthorized (line 125)
2. Access token expired (line 135) 
3. Cloudflare firewall kicked in (line 175) (so even if logged in, human can pass the check at least)

In Windows browsers when these 3 conditions do not exist, `https://chat.openai.com/api/auth/session` returns the following object:

```json
{"user":{"id":"user-9K0mEpF8oO9P6LsbYlV30ka6","name":"adamlui@protonmail.com","email":"adamlui@protonmail.com","image":"https://s.gravatar.com/avatar/3f6928f37e3fe9e016fed71c8c495608?s=480&r=pg&d=https%3A%2F%2Fcdn.auth0.com%2Favatars%2Fad.png","picture":"https://s.gravatar.com/avatar/3f6928f37e3fe9e016fed71c8c495608?s=480&r=pg&d=https%3A%2F%2Fcdn.auth0.com%2Favatars%2Fad.png","groups":[]},"expires":"2023-03-24T03:35:48.253Z","accessToken":"_______"}
```

Without the 'accessToken' value of this expected object, requests to `https://chat.openai.com/backend-api/conversation` (line 196) fail to generate a chat response. So based on your discovery, Apple browser is treating it like incognito & I don't know why :(

When I researched earlier the 1st error, I came across this answer: https://stackoverflow.com/a/59216403/2512339

"... led to this error on the iOS device emulator but not the Android device emulator. . Apple and Windows use different line endings, which is important considering template literals allow multiline strings"

Although no template literals are used, the answer does indicate something is different w/ how Apple handles requests (if that's of any help)

**Comment by adamlui** [comment]
> @adamlui This has nothing to do with browsers or extensions. Your script will only work in certain "normal" situations...

And what is abnormal to you? (not sure why the quotes)

>  ...because it lacks the necessary judgment (or error handling)

getAnswer() literally leads w/ a try-catch...

> For example you could check http-status, content-type... When the xhr request does not return a json type, you cannot use [JSON.parse](https://github.com/adamlui/userscripts/blob/7c7ce5948d7e81cf581617d937216a36706af4eb/chatgpt/duckduckgpt/duckduckgpt.user.js#L184-L196) to process the content.

`JSON.parse` in your linked getAccessToken() is in aforementioned error handling, so it doesn't matter if it's not json type (script will go to catch & clear token, then force user to login)

> Your [isBlockedbyCloudflare](https://github.com/adamlui/userscripts/blob/7c7ce5948d7e81cf581617d937216a36706af4eb/chatgpt/duckduckgpt/duckduckgpt.user.js#L95-L101) func is both unnecessary and very inaccurate

"Just a moment..." is the *only* situation where user is blocked by Cloudflare, so it is in fact *very* accurate. Why is it "unnecessary" for user to then be forced to clear it on login page? 

> on a 403 or not 200 status, you won't get "Just a moment..." every time.

Who said `isBlockedbyCloudflare` is for 403 or not 200 status? It's specifically for Cloudflare (it's literally in the name dude). If some other not-200 status shows, again, catch block of getAnswer() clears token & forces login (which Cloudflare checks don't require to lose token)

> If these code is not gpt's work, perhaps you can let it tell you how to write it. (forgive my joke :D)

No offense is taken, it works in every browser except Safari, why not put your money where your mouth is & create a commit?

**Comment by adamlui** [comment]
@ACTCD Surely the big game you talk means you can fix it in 5 minutes max. So don't be scared now, back your mouth up 😉 (or no one will believe you)

**Comment by quoid** [comment]
@adamlui 

> So based on your discovery, Apple browser is treating it like incognito & I don't know why :(

I should have been more clear. I got the userscript to pretty much work in Chrome/Violentmonkey on macOS. The issue I was seeing was specific to the Safari and even more specifically the background script of an extension. 

If I visit `https://chat.openai.com/api/auth/session` from a normal tab, I can see the token (expect response). But if I fetch that token from the background script it is an empty object 🤷 

My guess is that the background script context is a completely isolated world in Safari, so it is treat like an incognito tab. 

@ACTCD 

> This has nothing to do with browsers or extensions.

Speaking on the issue @Delamcode brings up, I am not entirely sure that is true. I can not be 100% sure because I only did some cursory exploration of the issue, but I think it has to do with the fact that background pages are an isolated environment. 

However, I **do think** that this is generally a good thing. I do not like the idea of background scripts being able to read auth tokens from users, for that I am not sure I would classify this as a bug, but perhaps a feature.

I am curious how ViolentMonkey seems to accomplish this...

Again, I would like to repeat that I did some hasty investigation so everything I say might not be accurate.

**Comment by adamlui** [comment]
> I should have been more clear. I got the userscript to pretty much work in Chrome/Violentmonkey on macOS. The issue I was seeing was specific to the Safari and even more specifically the background script of an extension.

Actually I said "Apple browser" but meant Safari & totally got what you meant!

Based on your detailed clarification though, I found a few posts of Safari WebExtensions' background requests always failing but not in Chrome (just as you described):

- CORS issue during HTTP request from background service worker in Safari web extension: https://stackoverflow.com/questions/73712933/cors-issue-during-http-request-from-background-service-worker-in-safari-web-exte

- Safari Web Extension background request CORS issues: https://developer.apple.com/forums/thread/654839

**Comment by Delamcode** [comment]
> @adamlui
> 
> > So based on your discovery, Apple browser is treating it like incognito & I don't know why :(
> 
> I should have been more clear. I got the userscript to pretty much work in Chrome/Violentmonkey on macOS. The issue I was seeing was specific to the Safari and even more specifically the background script of an extension.
> 
> If I visit `https://chat.openai.com/api/auth/session` from a normal tab, I can see the token (expect response). But if I fetch that token from the background script it is an empty object 🤷
> 
> My guess is that the background script context is a completely isolated world in Safari, so it is treated like an incognito tab.

Which is weird, because I have even tried disabling all privacy settings in safari and this is still happening… Could it be that extensions cannot communicate with other sites other than the one that is being used? Apple does after all sandbox apps on ios…

**Comment by quoid** [comment]
> Could it be that extensions cannot communicate with other sites other than the one that is being used? Apple does after all sandbox apps on ios…

@Delamcode I don't think that is it. You can fetch data from other APIs/sites without issue (example pasted below). I think it is just that the background page is isolated. The call `https://chat.openai.com/api/auth/session` is successful but it doesn't return the token because as far as it knows, you are not logged.


```js
// ==UserScript==
// @name        Example API call
// @description Example API call
// @match       *://*.example.com/*
// @grant       GM.xmlHttpRequest
// ==/UserScript==

(async () => {
    GM.xmlHttpRequest({
        method: "GET",
        url: "https://jsonplaceholder.typicode.com/posts",
        onloadend: e => {
            let response = e.response;
            try {
                response = JSON.parse(response);
            } catch (error) {
                console.error("Failed to parse JSON");
            }
            console.log(response);
        }
    });
})();
```

**Comment by Delamcode** [comment]
First, I don’t think this is possible, as the Userscript manager for ios (afaik) doesn’t have any settings. Also, it is not even showing up in the duckduckgo website as a running script :(

**Comment by adamlui** [comment]
@Delamcode Can you check safari console? Does an error like '`GM_registerMenuCommand` is not recognized' appear? If so I'm going to work on removing the menu and hard-code the proxy

**Comment by Delamcode** [comment]
Nope. No errors. It seems like the script didn’t grant duckduckgo.com?

**Comment by adamlui** [comment]
@Delamcode Oh I changed it to `// @match https://duckduckgo.com/?*` so it only appears on search results pages. But I just checked Safari Userscripts doesn't support menus, but I'm curious, what happens when you type a search?

**Comment by adamlui** [comment]
@Delamcode install this: https://github.com/kudoai/duckduckgpt/raw/main/greasemonkey/duckduckgpt-ios.user.js it should work

**Comment by Delamcode** [comment]
Tried that, and it showed up at least :) but error: ReferenceError {}

• column: 43
• line: 87
• message: "Can't find variable: GM_registerMenuCommand"
• sourceURL: "https://duckduckgo.com/?q=hello&t=ipad"
• ***@***.******@***.******@***.******@***.******@***.***://duckduckgo.com/?q=hello&t=ipad:181:15↵global ***@***.******@***.*** ***@***.******@***.******@***.***://872C2E0A-B9FE-4689-A249-28CC25A250F8/content.js:402:18"
• __proto__: Object

**Comment by adamlui** [comment]
@Delamcode `GM_registerMenuCommand` was removed from the updated script, so this error is coming from the old one (which always worked and showed on search results pages, just not on duckduckgo.com)

Can you try disabling this old script and install/enable the new one? (`duckduckgpt-ios.user.js`)

**Comment by Delamcode** [comment]
It seems the problem was actually another script (for that error). Now I don’t have any errors, but it still doesn’t load…

**Comment by adamlui** [comment]
@Delamcode what's the console showing for it?

**Comment by Delamcode** [comment]
Nothing, weirdly…

**Comment by adamlui** [comment]
@Delamcode I figured it out, the proxy I was using was limited to my own IP but I fixed it in this update: https://github.com/kudoai/duckduckgpt/blob/main/greasemonkey/duckduckgpt-ios.user.js can you pls try?

**Comment by Delamcode** [comment]
Still doesn’t work, and still no errors…

**Comment by adamlui** [comment]
Hey @Delamcode it works on Android now via Kiwi or Firefox which supports Tampermonkey, can you try intsalling [Firefox for iOS](https://apps.apple.com/us/app/firefox-private-safe-browser/id989804926) then [Tampermonkey for Firefox](https://addons.mozilla.org/en-US/firefox/addon/tampermonkey/) and confirm it works on iOS now too?

**Comment by Delamcode** [comment]
Firefox for iOS unfortunately doesn't support extensions. Another browser called Orion does, but I wasn't able to get tampermonkey/violentmonkey working.
